core/templates/pages/topics-and-skills-dashboard-page/topics-and-skills-dashboard-page.service.ts

```diff
@@ -0,0 +1,109 @@
+// Copyright 2020 The Oppia Authors. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS-IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/**
+ * @fileoverview Service for topics and skills dashboard page.
+ */
+
+import { downgradeInjectable } from '@angular/upgrade/static';
+import { Injectable } from '@angular/core';
+
+import { ITopicSummaryBackendDict } from
+  // eslint-disable-next-line max-len
+  'domain/topics_and_skills_dashboard/topics-and-skills-dashboard-backend-api.service';
+// eslint-disable-next-line max-len
+import { TopicsAndSkillsDashboardFilter } from
+  // eslint-disable-next-line max-len
+  'domain/topics_and_skills_dashboard/TopicsAndSkillsDashboardFilterObjectFactory';
+import { TopicsAndSkillsDashboardPageConstants } from
+  // eslint-disable-next-line max-len
+  'pages/topics-and-skills-dashboard-page/topics-and-skills-dashboard-page.constants';
+
+@Injectable({
+  providedIn: 'root'
+})
+export class TopicsAndSkillsDashboardPageService {
+  /**
+   * @param {Array} topicsArray - The original topics array
+   * @param {TopicsAndSkillsDashboardFilter} filterObject -
+   * the filter object values
+   * @returns {Array} filteredTopics - The filtered Topics array
+   */
+  getFilteredTopics(
+      topicsArray: Array<ITopicSummaryBackendDict>,
+      filterObject: TopicsAndSkillsDashboardFilter):
+      Array<ITopicSummaryBackendDict> {
+    let ESortOptions = TopicsAndSkillsDashboardPageConstants.TOPIC_SORT_OPTIONS;
+    let EPublishedOptions = (
+      TopicsAndSkillsDashboardPageConstants.TOPIC_PUBLISHED_OPTIONS);
+    let filteredTopics = topicsArray;
+    if (filterObject.keyword) {
+      filteredTopics = topicsArray.filter((topic) => {
+        return (
+          topic.name.toLowerCase().includes(
+            filterObject.keyword.toLowerCase()) ||
+            topic.description.toLowerCase().includes(
+              filterObject.keyword.toLowerCase()));
+      });
+    }
+
+    if (filterObject.category !== 'All') {
```

Comment:
Must use a named constant. (Never repeat raw strings/numbers throughout the codebase, for maintainability.)

Reply:
Done.